[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/howardddwu/notebook/blob/main/Assignment%203%20Equation%20of%20a%20Slime.ipynb)

## Haoyuan Wu
Assignemnt 3
Equation of a Slime

In [32]:
# libraries
import pandas as pd
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [33]:
# Using pandas load the dataset (load remotely, not locally)
dataframe = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
# Output the first 15 rows of the data
dataframe.head(15)


Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [34]:
# Display a summary of the table information (number of datapoints, etc.)
dataframe.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

## Part 2. Splitting the dataset

In [35]:
# Take the pandas dataset and split it into our features (X) and label (y)
x = dataframe[["Temperature °C", "Mols KCL"]]
y = dataframe["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=0)

## Part 3. Perform a Linear Regression

In [36]:
# Use sklearn to train a model on the training set
linear_regression = LinearRegression().fit(x_train,y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
x_sample_point = np.array([[200,420]])
y_sample_predict = linear_regression.predict([[200,420]])[0]
print(f"The predicted output from the model is {y_sample_predict}")
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
linear_train_score = linear_regression.score(x_train,y_train)
linear_test_score = linear_regression.score(x_test,y_test)
print(f"The linear train score is {train_score}")
print(f"The linear test score is {test_score}")


The predicted output from the model is 194983.82031779812
The linear train score is 0.8571870071495865
The linear test score is 0.8761646752736477


#### Explaination:
- The score mean the correlation between the predicted value vs. the actual value in the train dataset and the test dataset.
- The trian score of 0.857 meaning the predicted value from our model is relatively close to the actual value in the training set.<br>
- The test score of 0.876 meaning the predicted value from our model is relatively close to the actual value in the testing set.<br>
So that means our linear regression model is providing us a good result, since the best score is 1.

In [37]:
# Extract the coefficients and the intercept from the model and write an equation for your h(x) using LaTex
coef = linear_regression.coef_
coef[0] = round(coef[0],3)
coef[1] = round(coef[1],3)
intercept = round(linear_regression.intercept_,3)
print(f"The coefficients are x1 = {coef[0]}, x2 = {coef[1]}")
print(f"The intercept is {intercept}")

The coefficients are x1 = 863.581, x2 = 1006.127
The intercept is -400305.913


Equation: $y = 863.581x_{1} + 1006.127x_{2}-400305.913$
<br>Note: $x_{1}$ is Temperature, $x_{2}$ is Mols KCL

## Part 4. Use Cross Validation

In [38]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
fold_cv_10 = cross_val_score(linear_regression, x_train, y_train, cv = 10)
print(fold_cv_10)
print(f"The average of the 10-fold cross validation is {fold_cv_10.mean()}")

[0.82360363 0.84592328 0.80449234 0.88308699 0.87538513 0.87870759
 0.86953137 0.8583128  0.83288207 0.84650312]
The average of the 10-fold cross validation is 0.85184283230004


#### Report on their finding and their significance
- 10-fold cross validation is chosen meaning that the training set is divided to 10 small sets and each time use 9 sets to train and remaining one as testing set. Rotate the testing set to generate the test score.
- After doing that we have a average testing score as 0.8518 which is pretty good. Meaning that the model is good enough to predict the real world test set data.

## Part 5. Using Polynomial Regression

In [39]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly_regression = PolynomialFeatures(degree = 2)
poly_x_train = poly_regression.fit_transform(x_train)
poly_x_test = poly_regression.fit_transform(x_test)
poly_model = LinearRegression().fit(poly_x_train, y_train)

# Report on the metrics and output the resultant equation as you did in Part 3.
poly_train_score = poly_model.score(poly_x_train, y_train)
poly_test_score = poly_model.score(poly_x_test, y_test)
print(f"The polynomial train score is {poly_train_score}")
print(f"The polinomial test score is {poly_test_score}")

The polynomial train score is 1.0
The polinomial test score is 1.0


- The score of 1 means that the predicted value matches the value in the test set. Hence the polynomial regression model is a perfect model for this data.

In [40]:
# Extract the coefficients and the intercept from the model and write an equation for your h(x) using LaTex
coef = np.round_(poly_model.coef_,decimals=3)
intercept = round(poly_model.intercept_,3)
print(f"The coefficients are {coef}")
print(f"The intercept is {intercept}")

The coefficients are [ 0.    12.    -0.    -0.     2.     0.029]
The intercept is 0.0


Equation: $y = 12x_{1} + 2x_{1}x_{2}+0.029{x_{2}}^2$
<br>Note: $x_{1}$ is Temperature, $x_{2}$ is Mols KCL